src/mc-fle2-payload-iev.c

```diff
@@ -52,6 +55,230 @@ mc_FLE2IndexedEqualityEncryptedValueTokens_new (void)
+      (uint32_t) (sizeof (uint64_t) * 2 + sizeof (uint32_t) * 3),
+      &expected_plaintext_size,
+      status));
+
+   uint32_t expected_cipher_size = _mongocrypt_fle2_calculate_ciphertext_len (
+      expected_plaintext_size, status);
+   uint32_t expected_buf_size = 0;
+   CHECK_AND_RETURN (safe_uint32_t_sum (expected_cipher_size,
+                                        (uint32_t) (1 + sizeof (S_KeyId)),
+                                        &expected_buf_size,
+                                        status));
+
+   if (buf->len < expected_buf_size) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write buf is not large enough for iev");
+      return false;
+   }
+
+   mc_writer_t writer;
+   mc_writer_init_from_buffer (&writer, buf, __FUNCTION__);
+
+   const uint8_t subtype =
+      (uint8_t) MC_SUBTYPE_FLE2IndexedEqualityEncryptedValue;
+
+   if ((original_bson_type < 0) || (original_bson_type > 0xFF)) {
+      CLIENT_ERR ("Field 't' must be a valid BSON type, got: %d",
+                  original_bson_type);
+      return false;
+   }
+
+   const uint8_t bson_type = (uint8_t) original_bson_type;
+
+   mc_writer_write_u8 (&writer, subtype, status);
+   mc_writer_write_buffer (&writer, S_KeyId, S_KeyId->len, status);
+   mc_writer_write_u8 (&writer, bson_type, status);
+
+   mc_writer_write_buffer (
+      &writer, &encryption_out, encryption_out.len, status);
+
+   _mongocrypt_buffer_cleanup (&encryption_out);
+
+   return true;
+}
+
+#define CHECK_REMAINING_BUFFER_AND_RET(write_size)       \
+   if ((writer->pos + (write_size)) > writer->len) {     \
+      CLIENT_ERR ("%s expected byte "                    \
+                  "length >= %" PRIu64 " got: %" PRIu64, \
+                  writer->parser_name,                   \
+                  writer->pos + (write_size),            \
+                  writer->len);                          \
+      return false;                                      \
+   }
+
+static bool
+mc_fle2IndexedEncryptedValue_encrypt (
+   _mongocrypt_crypto_t *crypto,
+   const _mongocrypt_buffer_t *ClientEncryptedValue,
+   mc_ServerDataEncryptionLevel1Token_t *token,
+   mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
+   _mongocrypt_buffer_t *out,
+   mongocrypt_status_t *status)
+{
+#define CHECK_AND_GOTO(x) \
+   if (!(x)) {            \
+      goto cleanup;       \
+   }
+
+   bool ok = false;
+   _mongocrypt_buffer_t in;
+   _mongocrypt_buffer_t iv;
+
+   _mongocrypt_buffer_init (&in);
+   _mongocrypt_buffer_init_size (&iv, MONGOCRYPT_IV_LEN);
+
+   uint32_t expected_buf_size = 0;
+   CHECK_AND_GOTO (
+      safe_uint32_t_sum (ClientEncryptedValue->len,
+                         (uint32_t) (sizeof (uint64_t) * 2 + (32 * 3)),
+                         &expected_buf_size,
+                         status));
+
+   _mongocrypt_buffer_resize (&in, expected_buf_size);
+
+   uint32_t ciphertext_len =
+      _mongocrypt_fle2_calculate_ciphertext_len (expected_buf_size, status);
```

Comment:
Missing error handling on potential failure of `_mongocrypt_fle2_calculate_ciphertext_len`. If guaranteed to not fail, suggest adding a comment or `BSON_ASSERT()`.